client/my-sites/people/role-select/index.jsx

```diff
@@ -2,138 +2,87 @@
-
-	getInitialState() {
-		return ( {
-			roles: this.props.siteId ? RolesStore.getRoles( this.props.siteId ) : {}
-		} );
-	},
-
-	componentWillMount() {
-		if ( this.props.includeFollower ) {
-			this.refreshRoles();
-		}
-	},
-
-	componentDidMount() {
-		RolesStore.on( 'change', this.refreshRoles );
-		this.fetchRoles();
-	},
-
-	componentWillUnmount() {
-		RolesStore.removeListener( 'change', this.refreshRoles );
-	},
-
-	componentWillReceiveProps( nextProps ) {
-		const siteId = nextProps.siteId || this.props.siteId;
-		this.fetchRoles( siteId );
-		this.refreshRoles( nextProps );
-	},
-
-	getWPCOMFollowerRole( siteId ) {
-		siteId = siteId ? siteId : this.props.siteId;
-
-		const site = sites.getSite( siteId ),
-			displayName = site.is_private
-			? this.translate( 'Viewer', { context: 'Role that is displayed in a select' } )
-			: this.translate( 'Follower', { context: 'Role that is displayed in a select' } );
-
-		return {
-			follower: {
-				display_name: displayName,
-				name: 'follower'
-			}
-		};
-	},
-
-	refreshRoles( nextProps ) {
-		const siteId = nextProps && nextProps.siteId ? nextProps.siteId : this.props.siteId;
-
-		if ( siteId ) {
-			let siteRoles = RolesStore.getRoles( siteId );
-
-			if ( this.props.includeFollower ) {
-				siteRoles = Object.assign( {}, this.getWPCOMFollowerRole(), siteRoles );
-			}
-
-			this.setState( {
-				roles: siteRoles
-			} );
-		}
-	},
-
-	fetchRoles( siteId = this.props.siteId ) {
-		if ( ! siteId ) {
-			debug( 'siteId not set' );
-			return;
-		}
-
-		if ( Object.keys( RolesStore.getRoles( siteId ) ).length ) {
-			debug( 'initial fetch not necessary' );
-			return;
-		}
-
-		debug( 'Fetching roles for ' + siteId );
+import QuerySites from 'components/data/query-sites';
+import QuerySiteRoles from 'components/data/query-site-roles';
+import { getSite } from 'state/sites/selectors';
+import { getSiteRoles } from 'state/site-roles/selectors';
+
+const getWPCOMFollowerRole = ( { site, translate } ) => {
+	const displayName = site.is_private
+		? translate( 'Viewer', { context: 'Role that is displayed in a select' } )
+		: translate( 'Follower', { context: 'Role that is displayed in a select' } );
+
+	return {
+		display_name: displayName,
+		name: 'follower'
+	};
+};
+
+const RoleSelect = ( props ) => {
+	const { siteRoles, site, includeFollower, siteId, id, explanation, translate } = props;
+	const roles = siteRoles && siteRoles.slice( 0 );
+	const omitProps = [
+		'site',
+		'key',
+		'siteId',
+		'includeFollower',
+		'explanation',
+		'siteRoles',
+		'dispatch',
+		'moment',
+		'numberFormat',
+		'translate'
+	];
+
+	if ( site && roles && includeFollower ) {
+		roles.push( getWPCOMFollowerRole( props ) );
+	}
 
-		// defer fetch requests to avoid dispatcher conflicts
-		setTimeout( function() {
-			const fetching = RolesStore.isFetching( siteId );
-			if ( fetching ) {
-				return;
+	return (
+		<FormFieldset key={ siteId } disabled={ ! roles }>
+			{ siteId &&
```

Comment:
This won't work the way you think it does. The logic is such that it'll check the truthiness of each and stop at whichever is falsey or the final value, so it'd either render `null` for an empty site ID or the `<QuerySiteRoles />` component, but never the `<QuerySites />` component.

Instead, you might want:

```jsx
{ siteId && <QuerySites siteId={ siteId } /> }
{ siteId && <QuerySiteRoles siteId={ siteId } /> }
```

Or:

```jsx
{ siteId && [
	<QuerySites key="query-sites" siteId={ siteId } />,
	<QuerySiteRoles key="query-site-roles" siteId={ siteId } />
] }
```